text/0000-unicode-idents.md

```diff
@@ -0,0 +1,145 @@
+
+A new `confusable_unicode_idents` lint is added to the compiler. The default setting is `warn`.
+
+Note: The confusable detection is set to `warn` instead of `deny` to enable forward compatibility. The list of confusable characters will be extended in the future and programs that were once valid would fail to compile.
+
+The confusable detection algorithm is based on [Unicode® Technical Standard #39 Unicode Security Mechanisms Section 4 Confusable Detection][TR39Confusable]. For every distinct identifier X in the current scope execute the function `skeleton(X)`. If there exist two distinct identifiers X and Yin the same crate where `skeleton(X) = skeleton(Y)` report it.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* "ASCII is enough for anyone." As source code should be written in English and in English only (source: various people) no charactes outside the ASCII range are needed to express identifiers. Therefore support for Unicode identifiers introduces unnecceray complexity to the compiler.
+* "Foreign characters are hard to type." Usually computer keyboards provide access to the US-ASCII printable characters and the local language characters. Characters from other scripts are difficult to type, require entering numeric codes or are not available at all. These characters either need to be copy-pasted or entered with an alternative input method.
+* "Foreign characters are hard to read." If one is not familiar with the characters used it can be hard to tell them apart (e.g. φ and ψ) and one may not be able refer to the identifiers in an appropriate way (e.g. "loop" and "trident" instead of phi and psi)
+* Homoglyph attacks are possible. Without confusable detection identifiers can be distinct for the compiler but visually the same. Even with confusable detection there are still similar looking characters that may be confused by the casual reader.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+As stated in [Motivation](#motivation) allowing Unicode identifiers outside the ASCII range improves Rusts accessiability for developers not working in English. Especially in teaching and when the application domain vocabulary is not in English it can be beneficial to use names from the native language. To facilitate this it is necessary to allow a wide range of Unicode character in identifiers. The proposed implementation based on the Unicode TR31 is already used by other programming languages (e.g. Python 3) and is implemented behind the `non_ascii_idents` in *rustc* but lacks the NFKC normalization proposed.
+
+Possible variants:
+
+1. Require all identifiers to be in NFKC or NFC form.
+2. Two identifiers are only equal if their codepoints are equal.
+3. Perform NFC mapping instead of NFKC mapping for identifiers.
+4. Only a number of common scripts could be supported.
```

Comment:
Other possible variants are [the restriction levels of UTS #39](https://www.unicode.org/reports/tr39/#Restriction_Level_Detection).

Reply:
This is listed as variant 5.